File: src/main/scala/com/ajjpj/amapper/javabean/AnnotationBasedContextExtractor.scala
package com.ajjpj.amapper.javabean

import com.ajjpj.amapper.core.{AType, AContextExtractor}
import com.ajjpj.amapper.javabean.japi.AContextMarker

/**
 * registers an object as context value for all interfaces it implements that are annotated @AContextMarker
 *
 * @author arno
 */
class AnnotationBasedContextExtractor extends AContextExtractor {
  @volatile var cache = Map[AType, List[String]]()

  def withContext(context: Map[String, AnyRef], o: AnyRef, tpe: AType): Map[String, AnyRef] = {
    cache.get(tpe) match {
      case Some(Nil) =>
        context
      case Some(keys) =>
        context ++ keys.map(_ -> o).toMap
      case None =>
        val keys = extractKeys(tpe)
        cache += (tpe -> keys)
        context ++ keys.map(_ -> o).toMap
    }
  }

  def extractKeys(tpe: AType) = tpe match {
    //TODO recursive interface analysis; move this to JavaBeanSupport
    case SimpleJavaBeanType(cls) => cls.getInterfaces.filter(_.getAnnotation(classOf[AContextMarker]) != null).map(_.getName).toList
    case _ => Nil
  }
}

object AnnotationBasedContextExtractor extends AnnotationBasedContextExtractor

File: src/main/java/com/ajjpj/amapper/javabean/AnnotationBasedContextExtractor.java
package com.ajjpj.amapper.javabean;

import com.ajjpj.abase.collection.immutable.AList;
import com.ajjpj.abase.collection.immutable.AMap;
import com.ajjpj.amapper.core.AContextExtractor;
import com.ajjpj.amapper.core.tpe.AType;
import com.ajjpj.amapper.javabean.annotation.AContextMarker;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;


/**
 * @author arno
 */
public class AnnotationBasedContextExtractor implements AContextExtractor {
    private static final Map<AType, AList<String>> cache = new ConcurrentHashMap<AType, AList<String>>();

    @Override public AMap<String, Object> withContext(AMap<String, Object> context, Object o, AType tpe) {
        final AList<String> cacheValue = cache.get(tpe);
        if(cacheValue != null) {
            return withKeys(context, o, cacheValue);
        }

        final AList<String> keys = extractKeys(tpe);
        cache.put(tpe, keys);
        return withKeys(context, o, keys);
    }

    private AMap<String, Object> withKeys(AMap<String, Object> context, Object o, AList<String> keys) {
        AMap<String, Object> result = context;
        for(String key: keys) {
            result = result.updated(key, o);
        }
        return result;
    }

    private AList<String> extractKeys(AType tpe) {
        AList<String> result = AList.nil();

        if(tpe instanceof JavaBeanType<?>) {
            //TODO recursive interface analysis; move this to JavaBeanSupport
            for(Class<?> iface: ((JavaBeanType<?>) tpe).cls.getInterfaces()) {
                if(iface.getAnnotation(AContextMarker.class) != null) {
                    result = result.cons(iface.getName());
                }
            }
        }

        return result;
    }
}
